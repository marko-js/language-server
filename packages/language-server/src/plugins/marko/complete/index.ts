import { NodeType } from "@marko/language-tools";
import { CompletionItem } from "vscode-languageserver";

import { MarkoVirtualCode } from "../../../language";
import { AttrName } from "./AttrName";
import { Import } from "./Import";
import { OpenTagName } from "./OpenTagName";
import { Tag } from "./Tag";

export function provideCompletions(
  doc: MarkoVirtualCode,
  offset: number,
): CompletionItem[] | undefined {
  const node = doc.markoAst.nodeAt(offset);

  switch (node?.type) {
    case NodeType.AttrName:
      return AttrName(node, doc, offset);
    case NodeType.Import:
    case NodeType.Static:
      return Import(node, doc);
    case NodeType.Tag:
      return Tag(node, doc, offset);
    case NodeType.OpenTagName:
      return OpenTagName(node, doc);
    default:
      return;
  }
}
