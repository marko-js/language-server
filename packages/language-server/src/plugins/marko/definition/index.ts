import { NodeType } from "@marko/language-tools";
import { LocationLink } from "vscode-languageserver";

import { MarkoVirtualCode } from "../../../language";
import { AttrName } from "./AttrName";
import { OpenTagName } from "./OpenTagName";

export function provideDefinitions(
  doc: MarkoVirtualCode,
  offset: number,
): LocationLink[] | undefined {
  const node = doc.markoAst.nodeAt(offset);

  switch (node?.type) {
    case NodeType.AttrName:
      return AttrName(node, doc);
    case NodeType.OpenTagName:
      return OpenTagName(node, doc);
    default:
      return;
  }
}
