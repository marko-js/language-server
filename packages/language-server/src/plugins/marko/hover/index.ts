import { NodeType } from "@marko/language-tools";
import { Hover } from "vscode-languageserver";

import { MarkoVirtualCode } from "../../../language";
import { OpenTagName } from "./OpenTagName";

export function provideHover(
  doc: MarkoVirtualCode,
  offset: number,
): Hover | undefined {
  const node = doc.markoAst.nodeAt(offset);

  switch (node?.type) {
    case NodeType.OpenTagName:
      return OpenTagName(node, doc);
    default:
      return;
  }
}
