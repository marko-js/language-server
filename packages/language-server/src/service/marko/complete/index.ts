import type { CompletionItem, CompletionParams } from "vscode-languageserver";
import { NodeType } from "@marko/language-tools";

import { MarkoFile, getMarkoFile } from "../../../utils/file";
import type { Plugin, Result } from "../../types";

import { AttrName } from "./AttrName";
import { AttrValue } from "./AttrValue";
import { Import } from "./Import";
import { OpenTagName } from "./OpenTagName";
import { Tag } from "./Tag";

export type CompletionResult = Result<CompletionItem[]>;
export interface CompletionMeta<N = unknown> {
  file: MarkoFile;
  params: CompletionParams;
  offset: number;
  node: N;
}

const handlers: Record<
  string,
  (data: CompletionMeta<any>) => CompletionResult
> = {
  Tag,
  OpenTagName,
  AttrName,
  AttrValue,
  Import,
};

export const doComplete: Plugin["doComplete"] = async (doc, params) => {
  const file = getMarkoFile(doc);
  const offset = doc.offsetAt(params.position);
  const node = file.parsed.nodeAt(offset);
  return {
    items:
      (await handlers[NodeType[node.type]]?.({
        file,
        params,
        offset,
        node,
      })) || [],
    isIncomplete: true,
  };
};
