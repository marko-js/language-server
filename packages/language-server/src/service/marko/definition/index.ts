import type { DefinitionLink, DefinitionParams } from "vscode-languageserver";
import { NodeType } from "@marko/language-tools";

import { MarkoFile, getMarkoFile } from "../../../utils/file";
import type { Plugin, Result } from "../../types";

import { AttrName } from "./AttrName";
import { OpenTagName } from "./OpenTagName";

export type DefinitionResult = Result<DefinitionLink[]>;
export interface DefinitionMeta<N = unknown> {
  file: MarkoFile;
  params: DefinitionParams;
  offset: number;
  node: N;
}

const handlers: Record<
  string,
  (data: DefinitionMeta<any>) => DefinitionResult
> = {
  OpenTagName,
  AttrName,
};

export const findDefinition: Plugin["findDefinition"] = async (doc, params) => {
  const file = getMarkoFile(doc);
  const offset = doc.offsetAt(params.position);
  const node = file.parsed.nodeAt(offset);
  return (
    (await handlers[NodeType[node.type]]?.({
      file,
      params,
      offset,
      node,
    })) || []
  );
};
