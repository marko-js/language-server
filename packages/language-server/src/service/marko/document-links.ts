import type { DocumentLink } from "vscode-languageserver";
import { type Node, NodeType } from "@marko/language-tools";
import { MarkoFile, processDoc } from "../../utils/file";

import resolveUrl from "../../utils/resolve-url";
import type { Plugin } from "../types";

import isDocumentLinkAttr from "./util/is-document-link-attr";

const importTagReg = /(['"])<((?:[^\1\\>]+|\\.)*)>?\1/g;

export const findDocumentLinks: Plugin["findDocumentLinks"] = async (doc) => {
  return processDoc(doc, extractDocumentLinks);
};

/**
 * Iterate over the Marko CST and extract all the file links in the document.
 */
function extractDocumentLinks({
  uri,
  scheme,
  parsed,
  code,
  project: { lookup },
}: MarkoFile): DocumentLink[] {
  if (scheme !== "file") {
    return [];
  }

  const links: DocumentLink[] = [];
  const { program, read } = parsed;
  const visit = (node: Node.ChildNode) => {
    switch (node.type) {
      case NodeType.AttrTag:
        if (node.body) {
          for (const child of node.body) {
            visit(child);
          }
        }
        break;
      case NodeType.Tag:
        if (node.attrs && node.nameText) {
          for (const attr of node.attrs) {
            if (isDocumentLinkAttr(code, node, attr)) {
              const resolved = resolveUrl(
                read(attr.value.value).slice(1, -1),
                uri
              );
              if (resolved) {
                links.push({
                  range: parsed.locationAt(attr.value.value),
                  target: resolveUrl(read(attr.value.value).slice(1, -1), uri),
                });
              }
            }
          }
        }
        if (node.body) {
          for (const child of node.body) {
            visit(child);
          }
        }

        break;
    }
  };

  for (const node of program.static) {
    // check for import statement (this currently only support the tag import shorthand).
    if (node.type === NodeType.Import) {
      importTagReg.lastIndex = 0;
      const value = parsed.read(node);
      const match = importTagReg.exec(value);
      if (match) {
        const [{ length }, , tagName] = match;
        const tagDef = lookup.getTag(tagName);
        const fileForTag = tagDef && (tagDef.template || tagDef.renderer);

        if (fileForTag) {
          links.push({
            range: parsed.locationAt({
              start: node.start + match.index,
              end: node.start + match.index + length,
            }),
            target: fileForTag,
          });
        }
      }
    }
  }

  for (const node of program.body) {
    visit(node);
  }

  return links;
}
