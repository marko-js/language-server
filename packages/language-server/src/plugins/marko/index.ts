import {
  LanguageServicePlugin,
  LanguageServicePluginInstance,
} from "@volar/language-service";
import { TextDocument } from "vscode-languageserver-textdocument";
import { URI } from "vscode-uri";

import { MarkoVirtualCode } from "../../language";
import { provideCompletions } from "./complete";
import { provideDefinitions } from "./definition";
import { provideHover } from "./hover";
import { provideValidations } from "./validate";
// import { provideDocumentSymbols } from "./document-symbols";

export const create = (
  _: typeof import("typescript"),
): LanguageServicePlugin => {
  return {
    name: "marko",
    capabilities: {
      hoverProvider: true,
      definitionProvider: true,
      diagnosticProvider: {
        interFileDependencies: false,
        workspaceDiagnostics: false,
      },
      completionProvider: {
        triggerCharacters: [
          ".",
          ":",
          "<",
          ">",
          "@",
          "/",
          '"',
          "'",
          "`",
          " ",
          "=",
          "*",
          "#",
          "$",
          "+",
          "^",
          "(",
          "[",
          "-",
        ],
      },
    },
    create(context): LanguageServicePluginInstance {
      return {
        // TODO: Is this necessary?
        // provideDocumentSymbols(document, token) {
        //   if (token.isCancellationRequested) return;
        //   return worker(document, (virtualCode) => {
        //     return provideDocumentSymbols(virtualCode);
        //   });
        // },
        provideDefinition(document, position, token) {
          if (token.isCancellationRequested) return;
          return worker(document, (virtualCode) => {
            const offset = document.offsetAt(position);
            return provideDefinitions(virtualCode, offset);
          });
        },
        provideDiagnostics(document, token) {
          if (token.isCancellationRequested) return;
          return worker(document, async (virtualCode) => {
            return await provideValidations(virtualCode);
          });
        },
        provideHover(document, position, token) {
          if (token.isCancellationRequested) return;
          return worker(document, (virtualCode) => {
            const offset = document.offsetAt(position);
            return provideHover(virtualCode, offset);
          });
        },
        provideCompletionItems(document, position, _, token) {
          if (token.isCancellationRequested) return;
          return worker(document, (virtualCode) => {
            const offset = document.offsetAt(position);
            const completions = provideCompletions(virtualCode, offset);

            if (completions) {
              return {
                isIncomplete: false,
                items: completions.map((it) => {
                  it.data = { source: "marko" };
                  return it;
                }),
              };
            }

            return {
              items: [],
              isIncomplete: true,
            };
          });
        },
      };

      function worker<T>(
        document: TextDocument,
        callback: (markoDocument: MarkoVirtualCode) => T,
      ): T | undefined {
        const decoded = context.decodeEmbeddedDocumentUri(
          URI.parse(document.uri),
        );
        const sourceScript =
          decoded && context.language.scripts.get(decoded[0]);
        const virtualCode =
          decoded && sourceScript?.generated?.embeddedCodes.get(decoded[1]);
        if (!(virtualCode instanceof MarkoVirtualCode)) return;

        return callback(virtualCode);
      }
    },
  };
};
