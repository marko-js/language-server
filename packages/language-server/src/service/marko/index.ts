import type { Plugin } from "../types";

import { doComplete } from "./complete";
import { doValidate } from "./validate";
import { doHover } from "./hover";
import { findDefinition } from "./definition";
import { findDocumentLinks } from "./document-links";
import { findDocumentSymbols } from "./document-symbols";
import { format } from "./format";

export default {
  doComplete,
  doValidate,
  doHover,
  findDefinition,
  findDocumentLinks,
  findDocumentSymbols,
  format,
} as Plugin;
