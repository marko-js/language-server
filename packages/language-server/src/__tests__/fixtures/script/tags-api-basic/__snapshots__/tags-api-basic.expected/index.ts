import FancyButton from "./components/fancy-button/index.marko";
import { get } from "@ebay/retriever";
/** Hi */ export interface Input {
  year: number;
  isSmartOnly: boolean;
  type: string;
  mobileList: Array<Mobile>;
  renderBody: Marko.Body;
}

function greet() {
  return "hello world";
}

var useA = true;
abstract class Component extends Marko.Component<Input> {}
export { type Component };
function __marko_internal_template(this: void) {
  const input = Marko._.any as Input;
  const component = Marko._.any as Component;
  const state = Marko._.state(component);
  const out = Marko._.out;
  const $signal = Marko._.any as AbortSignal;
  const $global = Marko._.getGlobal(
    // @ts-expect-error We expect the compiler to error because we are checking if the MarkoRun.Context is defined.
    (Marko._.error, Marko._.any as MarkoRun.Context),
  );
  Marko._.noop({ component, state, out, input, $global, $signal });
  const __marko_internal_tag_1 = Marko._.resolveTemplate(
    import("../../../components/const/index.marko"),
  );
  Marko._.assertRendered(
    Marko._.rendered,
    1,
    Marko._.renderTemplate(__marko_internal_tag_1)()()({
      value: input,
    }),
  );
  const { year, isSmartOnly, type, mobileList, renderBody } =
    Marko._.rendered.returns[1].value;
  Marko._.renderNativeTag("div")()()({
    class: "mobiles__list",
    ["renderBody" /*div*/]: (() => {
      Marko._.renderNativeTag("p")()()({
        id: "p",
        ["renderBody" /*p*/]: (() => {
          year;
          return () => {
            return Marko._.voidReturn;
          };
        })(),
      });
      if (isSmartOnly) {
        Marko._.renderNativeTag("span")()()({
          class: "subnote",
          ["renderBody" /*span*/]: (() => {
            isSmartOnly;
            return () => {
              return Marko._.voidReturn;
            };
          })(),
        });
      }
      Marko._.renderNativeTag("span")()()({
        class: "subnote",
        ["renderBody" /*span*/]: (() => {
          type;
          return () => {
            return Marko._.voidReturn;
          };
        })(),
      });
      Marko._.forOfTag(
        {
          /*for*/ of: mobileList,
        },
        (mobile, i, all) => {
          Marko._.renderNativeTag("span")()()(
            //                          ^?
            {
              id: i,
              ["renderBody" /*span*/]: (() => {
                mobile;
                i;
                all.length;
                return () => {
                  return Marko._.voidReturn;
                };
              })(),
            },
          );
          return Marko._.voidReturn;
        },
      );
      Marko._.renderNativeTag("div")()()(
        //                 ^?        ^?
        {
          class: Marko._.interpolated`container`,
          ["renderBody" /*div*/]: (() => {
            const __marko_internal_tag_2 = renderBody;
            Marko._.renderDynamicTag(__marko_internal_tag_2)()()({
              /*renderBody*/
            });
            return () => {
              return Marko._.voidReturn;
            };
          })(),
        },
      );
      const __marko_internal_tag_3 = FancyButton;
      Marko._.renderDynamicTag(__marko_internal_tag_3)()()({
        something: true,
        ["renderBody" /*FancyButton*/]: (() => {
          type;
          return () => {
            return Marko._.voidReturn;
          };
        })(),
      });
      Marko._.renderNativeTag("div")()()({
        ["renderBody" /*div*/]: (() => {
          ("<b>World</b>");
          return () => {
            return Marko._.voidReturn;
          };
        })(),
      });
      Marko._.renderNativeTag("div")()()({
        ["renderBody" /*div*/]: (() => {
          Marko._.renderNativeTag("code")()()({
            ["renderBody" /*code*/]: (() => {
              return () => {
                return Marko._.voidReturn;
              };
            })(),
          });
          return () => {
            return Marko._.voidReturn;
          };
        })(),
      });
      return () => {
        return Marko._.voidReturn;
      };
    })(),
  });
  const __marko_internal_return = {
    return: Marko._.returnTag({
      value: 1,
    }),
  };
  return __marko_internal_return.return;
}
export default new (class Template extends Marko._.Template<{
  render(
    input: Marko.TemplateInput<Input>,
    stream?: {
      write: (chunk: string) => void;
      end: (chunk?: string) => void;
    },
  ): Marko.Out<Component>;

  render(
    input: Marko.TemplateInput<Input>,
    cb?: (err: Error | null, result: Marko.RenderResult<Component>) => void,
  ): Marko.Out<Component>;

  renderSync(input: Marko.TemplateInput<Input>): Marko.RenderResult<Component>;

  renderToString(input: Marko.TemplateInput<Input>): string;

  stream(
    input: Marko.TemplateInput<Input>,
  ): ReadableStream<string> & NodeJS.ReadableStream;

  api: "class";
  _(): () => <__marko_internal_input extends unknown>(
    input: Marko.Directives &
      Input &
      Marko._.Relate<__marko_internal_input, Marko.Directives & Input>,
  ) => Marko._.ReturnWithScope<
    __marko_internal_input,
    typeof __marko_internal_template extends () => infer Return ? Return : never
  >;
}> {})();
