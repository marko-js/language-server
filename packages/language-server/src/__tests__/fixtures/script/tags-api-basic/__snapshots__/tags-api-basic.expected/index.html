<div data-marko-node-id="1" class="mobiles__list">
    <p data-marko-node-id="2" id="p">Year: placeholder</p>

    <div>
      <span data-marko-node-id="4" class="subnote">isSmartPhone: placeholder</span>
    </div>

    <span data-marko-node-id="5" class="subnote">phoneType: placeholder</span>

    <div>

      <span data-marko-node-id="7" id="dynamic">placeholder placeholder placeholder</span>

    </div>

    <div data-marko-node-id="8">
        
    </div>
    <div>placeholder</div>
    <div data-marko-node-id="11">Hello placeholder</div>
    <div data-marko-node-id="12">Placeholder example: <code data-marko-node-id="13">${someValue}</code></div>
</div>